src/NUnitEngine/nunit.engine.core/Drivers/NUnitFrameworkApi2018.cs

```diff
@@ -0,0 +1,253 @@
+            }
+            catch (Exception e)
+            {
+                var msg = string.Format(FAILED_TO_LOAD_ASSEMBLY + assemblyPath);
+                log.Error(msg);
+                throw new NUnitEngineException(msg, e);
+            }
+
+            log.Debug($"Loaded {assemblyPath}");
+            return assembly;
+        }
+
+        private Assembly LoadAssembly(AssemblyName assemblyName)
+        {
+            Assembly assembly;
+
+            try
+            {
+                assembly = _assemblyLoadContext?.LoadFromAssemblyName(assemblyName)!;
+                if (assembly == null)
+                    throw new Exception("LoadFromAssemblyName returned null");
+            }
+            catch (Exception e)
+            {
+                var msg = string.Format(FAILED_TO_LOAD_ASSEMBLY + assemblyName.FullName);
+                log.Error($"{FAILED_TO_LOAD_ASSEMBLY}\r\n{e}");
+                throw new NUnitEngineException(FAILED_TO_LOAD_NUNIT, e);
+            }
+
+            log.Debug($"Loaded {assemblyName.FullName}");
+            return assembly;
+        }
+#endif
+
+        // API methods with no overloads
+        private static readonly string LOAD_METHOD = "LoadTests";
+        private static readonly string EXPLORE_METHOD = "ExploreTests";
+        private static readonly string COUNT_METHOD = "CountTests";
+        private static readonly string STOP_RUN_METHOD = "StopRun";
+
+        // Execute methods with no overloads
+        private object ExecuteMethod(string methodName, params object?[] args)
+        {
+            log.Debug($"Calling API method {methodName} with args {string.Join("+", args)}");
+            var method = _frameworkControllerType.ShouldNotBeNull().GetMethod(methodName, BindingFlags.Public | BindingFlags.Instance);
+            return ExecuteMethod(method, args);
+        }
+
+        // API methods with overloads
+        private static readonly string RUN_METHOD = "RunTests";
+        private static readonly string RUN_ASYNC_METHOD = "RunTests";
+
+        // Execute overloaded methods specifying argument types
+        private object ExecuteMethod(string methodName, Type[] ptypes, params object?[] args)
+        {
+            log.Debug($"Calling API method {methodName} with arg types {string.Join<Type>("+", ptypes)}");
+            var method = _frameworkControllerType.ShouldNotBeNull().GetMethod(methodName, ptypes);
+            return ExecuteMethod(method, args);
+        }
+
+        private object ExecuteMethod(MethodInfo? method, params object?[] args)
+        {
+            if (method == null)
+                throw new NUnitEngineException(INVALID_FRAMEWORK_MESSAGE);
+
+            log.Debug($"Executing {method.DeclaringType}.{method.Name}");
+
+#if NETFRAMEWORK
+            return method.Invoke(_frameworkController, args).ShouldNotBeNull();
+#else
+            using (_assemblyLoadContext.ShouldNotBeNull().EnterContextualReflection())
```

Comment:
If I interpret the [documentation](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/AssemblyLoadContext.ContextualReflection.md) correctly this is not needed.
The context needs setting when the instance is created, not when one of it's methods is called.

Reply:
This is code from the original driver that I simply moved to the api class. To be honest, I didn't understand why it was needed but I hesitated to change it. I'll try removing and leave it out if nothing breaks.

Reply:
You're right. Not needed. Doesn't seem to be needed when creating the instance either.

Reply:
I hadn't heard of `AssemblyLoadContext` before.
If I interpreted the referenced documentation correctly, you can remove it here as it wouldn't be used.
That is _unless_ the test (controller) loads something indirect via reflection, e.g. creating a test fixture instance.

I think it might be important when the runner/engine and the testassembly both depend on a common assembly but use different versions. The test should load the version the testassembly wants, not the one the runner uses.
I will see if I can make some completely separate test code to see if it matters.

Reply:
In case it helps, here's history... @rprouse wrote this for .NET Standard, where there isn't any `AssemblyLoadContext`. @ChrisMaddock  (I believe) enhanced it for .NET 3.1, which introduced `AssemblyLoadContext` and made it easier to take charge of how tests were loaded in a specific context. I don't think it has had much change since then.

In any case, I'd like to get this in so we can look at some other stuff. If you come up with something different, let's just do a bug or refactoring issue.